src/mds/MetricsHandler.cc

```diff
@@ -0,0 +1,241 @@
+  // update this rank.
+  auto lus = it->second.first;
+  if (last_updated_seq == 0 || lus == last_updated_seq) {
+    dout(10) << ": metric lus=" << lus << ", last_updated_seq=" << last_updated_seq
+             << dendl;
+    client_metrics_map.erase(it);
+    return;
+  }
+
+  // zero out all metrics
+  auto &metrics = it->second.second;
+  metrics.cap_hit_metric = { };
+  metrics.read_latency_metric = { };
+  metrics.write_latency_metric = { };
+  metrics.metadata_latency_metric = { };
+  metrics.update_type = UPDATE_TYPE_REMOVE;
+}
+
+void MetricsHandler::set_next_seq(version_t  seq) {
+  dout(10) << ": current sequence number " << next_seq << ", setting next sequence number "
+           << seq << dendl;
+  last_updated_seq = next_seq = seq;
+}
+
+void MetricsHandler::handle_payload(Session *session, const CapInfoPayload &payload) {
+  dout(20) << ": session=" << session << ", hits=" << payload.cap_hits << ", misses="
+           << payload.cap_misses << dendl;
+
+  auto it = client_metrics_map.find(session->info.inst);
+  if (it == client_metrics_map.end()) {
+    return;
+  }
+
+  auto &metrics = it->second.second;
+  metrics.update_type = UPDATE_TYPE_REFRESH;
+  metrics.cap_hit_metric.hits = payload.cap_hits;
+  metrics.cap_hit_metric.misses = payload.cap_misses;
+}
+
+void MetricsHandler::handle_payload(Session *session, const ReadLatencyPayload &payload) {
+  dout(20) << ": session=" << session << ", latency=" << payload.lat << dendl;
+
+  auto it = client_metrics_map.find(session->info.inst);
+  if (it == client_metrics_map.end()) {
+    return;
+  }
+
+  auto &metrics = it->second.second;
+  metrics.update_type = UPDATE_TYPE_REFRESH;
+  metrics.read_latency_metric.lat = payload.lat;
+}
+
+void MetricsHandler::handle_payload(Session *session, const WriteLatencyPayload &payload) {
+  dout(20) << ": session=" << session << ", latency=" << payload.lat << dendl;
+
+  auto it = client_metrics_map.find(session->info.inst);
+  if (it == client_metrics_map.end()) {
+    return;
+  }
+
+  auto &metrics = it->second.second;
+  metrics.update_type = UPDATE_TYPE_REFRESH;
+  metrics.write_latency_metric.lat = payload.lat;
+}
+
+void MetricsHandler::handle_payload(Session *session, const MetadataLatencyPayload &payload) {
+  dout(20) << ": session=" << session << ", latency=" << payload.lat << dendl;
+
+  auto it = client_metrics_map.find(session->info.inst);
+  if (it == client_metrics_map.end()) {
+    return;
+  }
+
+  auto &metrics = it->second.second;
+  metrics.update_type = UPDATE_TYPE_REFRESH;
+  metrics.metadata_latency_metric.lat = payload.lat;
+}
+
+void MetricsHandler::handle_payload(Session *session, const UnknownPayload &payload) {
+  dout(5) << ": session=" << session << ", ignoring unknown payload" << dendl;
+}
+
+void MetricsHandler::handle_client_metrics(const cref_t<MClientMetrics> &m) {
+  std::scoped_lock locker(lock);
+
+  Session *session = mds->get_session(m);
+  dout(20) << __func__ << ": session=" << session << dendl;
```

Comment:
We could remove all the __func__ , it is duplicated.

```
out/mds.c.log:81256:2019-12-19T06:19:23.530-0500 7fabf99d0700 20 handle_client_metrics: mds.metricshandle_client_metrics: session=0x560d23892880
out/mds.c.log:81626:2019-12-19T06:19:29.162-0500 7fabf99d0700 20 handle_client_metrics: mds.metricshandle_client_metrics: session=0x560d23892880
out/mds.c.log:82074:2019-12-19T06:19:34.282-0500 7fabf99d0700 20 handle_client_metrics: mds.metricshandle_client_metrics: session=0x560d23892880
out/mds.c.log:82462:2019-12-19T06:19:39.401-0500 7fabf99d0700 20 handle_client_metrics: mds.metricshandle_client_metrics: session=0x560d23892880
out/mds.c.log:82951:2019-12-19T06:19:45.545-0500 7fabf99d0700 20 handle_client_metrics: mds.metricshandle_client_metrics: session=0x560d23892880
out/mds.c.log:83327:2019-12-19T06:19:51.178-0500 7fabf99d0700 20 handle_client_metrics: mds.metricshandle_client_metrics: session=0x560d23892880
out/mds.c.log:83804:2019-12-19T06:19:56.297-0500 7fabf99d0700 20 handle_client_metrics: mds.metricshandle_client_metrics: session=0x560d23892880
out/mds.c.log:84137:2019-12-19T06:20:01.418-0500 7fabf99d0700 20 handle_client_metrics: mds.metricshandle_client_metrics: session=0x560d23892880
out/mds.c.log:84650:2019-12-19T06:20:07.567-0500 7fabf99d0700 20 handle_client_metrics: mds.metricshandle_client_metrics: session=0x560d23892880
out/mds.c.log:84992:2019-12-19T06:20:13.193-0500 7fabf99d0700 20 handle_client_metrics: mds.metricshandle_client_metrics: session=0x560d23892880
out/mds.c.log:85466:2019-12-19T06:20:18.314-0500 7fabf99d0700 20 handle_client_metrics: mds.metricshandle_client_metrics: session=0x560d23892880
out/mds.c.log:85821:2019-12-19T06:20:23.432-0500 7fabf99d0700 20 handle_client_metrics: mds.metricshandle_client_metrics: session=0x560d23892880

```

Reply:
ack -- I'll do the changes.

Reply:
`__func__` is duplicated in couple of places in this source. cleanup up that.

Reply:
done